File: src/main/java/io/zeitwert/ddd/property/model/impl/EnumSetPropertyImpl.java
package io.zeitwert.ddd.property.model.impl;

import io.zeitwert.ddd.enums.model.Enumerated;
import io.zeitwert.ddd.enums.model.Enumeration;
import io.zeitwert.ddd.part.model.base.PartSPI;
import io.zeitwert.ddd.part.model.enums.CodePartListType;
import io.zeitwert.ddd.property.model.EnumSetProperty;
import io.zeitwert.ddd.property.model.AggregatePartItem;
import io.zeitwert.ddd.property.model.base.EntityWithPropertiesSPI;
import io.zeitwert.ddd.property.model.base.PropertyBase;

import java.util.Collection;
import java.util.HashSet;
import java.util.Set;

import static io.zeitwert.ddd.util.Check.assertThis;

public class EnumSetPropertyImpl<E extends Enumerated> extends PropertyBase<E> implements EnumSetProperty<E> {

	private final String name;
	private final CodePartListType partListType;
	private final Enumeration<E> enumeration;

	private Set<AggregatePartItem<?>> itemSet = new HashSet<>();

	public EnumSetPropertyImpl(EntityWithPropertiesSPI entity, CodePartListType partListType,
			Enumeration<E> enumeration) {
		super(entity);
		this.name = partListType.getId();
		this.partListType = partListType;
		this.enumeration = enumeration;
	}

	public EnumSetPropertyImpl(EntityWithPropertiesSPI entity, String name, CodePartListType partListType,
			Enumeration<E> enumeration) {
		super(entity);
		this.name = name;
		this.partListType = partListType;
		this.enumeration = enumeration;
	}

	@Override
	public String getName() {
		return this.name;
	}

	@Override
	public CodePartListType getPartListType() {
		return this.partListType;
	}

	@Override
	public void clearItems() {
		this.itemSet.forEach(item -> ((PartSPI<?>) item).delete());
		this.itemSet.clear();
		this.getEntity().afterClear(this);
	}

	@Override
	public void addItem(E item) {
		assertThis(item != null, "item not null");
		if (item == null) {
			return; // make compiler happy (potential null pointer)
		}
		if (!this.hasItem(item)) {
			AggregatePartItem<?> part = (AggregatePartItem<?>) this.getEntity().addPart(this, this.partListType);
			assertThis(part != null,
					"entity " + this.getEntity().getClass().getSimpleName() + "created a part for " + this.partListType.getId()
							+ " (make sure to compare property with .equals() in addPart)");
			if (part == null) {
				return; // make compiler happy (potential null pointer)
			}
			part.setItemId(item.getId());
			this.itemSet.add(part);
			this.getEntity().afterAdd(this);
		}
	}

	@Override
	public Set<E> getItems() {
		return Set.copyOf(this.itemSet.stream().map(item -> this.enumeration.getItem(item.getItemId())).toList());
	}

	@Override
	public boolean hasItem(E item) {
		for (AggregatePartItem<?> part : this.itemSet) {
			if (part.getItemId().equals(item.getId())) {
				return true;
			}
		}
		return false;
	}

	@Override
	public void removeItem(E item) {
		assertThis(item != null, "item not null");
		if (item == null) {
			return; // make compiler happy (potential null pointer)
		}
		if (this.hasItem(item)) {
			AggregatePartItem<?> part = this.itemSet.stream().filter(p -> p.getItemId().equals(item.getId())).findAny().get();
			((PartSPI<?>) part).delete();
			this.itemSet.remove(part);
			this.getEntity().afterRemove(this);
		}
	}

	public void doBeforeStore() {
		int seqNr = 0;
		for (AggregatePartItem<?> item : this.itemSet) {
			item.setSeqNr(seqNr++);
		}
	}

	@Override
	public void loadEnums(Collection<? extends AggregatePartItem<?>> partList) {
		this.itemSet.clear();
		partList.forEach(p -> this.itemSet.add(p));
	}

}

File: fm-dddrive/src/main/java/dddrive/ddd/property/model/impl/EnumSetPropertyImpl.kt
package dddrive.ddd.property.model.impl

import dddrive.ddd.enums.model.Enumerated
import dddrive.ddd.enums.model.Enumeration
import dddrive.ddd.property.model.EntityWithProperties
import dddrive.ddd.property.model.EntityWithPropertiesSPI
import dddrive.ddd.property.model.EnumSetProperty
import dddrive.ddd.property.model.base.PropertyBase

class EnumSetPropertyImpl<E : Enumerated>(
	entity: EntityWithProperties,
	name: String,
	override val enumeration: Enumeration<E>,
) : PropertyBase<E>(entity, name),
	EnumSetProperty<E> {

	private val items: MutableSet<E> = mutableSetOf()

	override fun clear() {
		require(isWritable) { "writable" }
		items.toList().forEach { remove(it) }
		check(items.isEmpty()) { "items empty" }
		(entity as EntityWithPropertiesSPI).doAfterClear(this)
	}

	override fun add(item: E) {
		require(isWritable) { "writable" }
		require(isValidEnum(item)) { "valid enumeration item for " + enumeration.id + " (" + item.id + ")" }
		if (!has(item)) {
			val entity = entity as EntityWithPropertiesSPI
			fireValueAddedChange(item.id)
			items.add(item)
			entity.doAfterAdd(this, null)
		}
	}

	override fun has(item: E): Boolean = items.contains(item)

	override fun remove(item: E) {
		require(isWritable) { "writable" }
		if (has(item)) {
			val entity = entity as EntityWithPropertiesSPI
			fireValueRemovedChange(item.id)
			items.remove(item)
			entity.doAfterRemove(this)
		}
	}

	private fun isValidEnum(value: E): Boolean {
		if (value.enumeration != enumeration) {
			return false
		}
		return value == enumeration.getItem(value.id)
	}

	override val size: Int get() = items.size

	override fun isEmpty(): Boolean = items.isEmpty()

	override fun iterator(): Iterator<E> = items.iterator()

	override fun containsAll(elements: Collection<E>): Boolean = items.containsAll(elements)

	override fun contains(element: E): Boolean = items.contains(element)

}
